cmd/processor/app/processor.go

```diff
@@ -649,3 +654,32 @@ func (p *Processor) setWorkersStatus(triggerInstance trigger.Trigger, status sta
 
 	return nil
 }
+
+// handleSignals creates a signal handler, so on signal processor gracefully terminated
+func (p *Processor) handleSignals() {
+	var captureSignal = make(chan os.Signal, 1)
+
+	// when k8s deletes pods, it sends SIGTERM (https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)
+	// when docker stops container it sends SIGTERM by default (https://docs.docker.com/engine/reference/commandline/stop/)
+	// but you can specify a specific signal with a specific option, so we support SIGABRT and SIGINT as well
+	signal.Notify(captureSignal, syscall.SIGTERM, syscall.SIGABRT, syscall.SIGINT)
+	p.terminateAllTriggers(<-captureSignal)
+	p.Stop()
+}
+
+func (p *Processor) terminateAllTriggers(signal os.Signal) {
+	p.logger.WarnWith("Got system signal", "signal", signal.String())
+
+	wg := &sync.WaitGroup{}
+	for _, triggerInstance := range p.triggers {
+		wg.Add(1)
+
+		// drains all workers in trigger (for each trigger in parallel)
+		go func(triggerInstance trigger.Trigger, wg *sync.WaitGroup) {
+			defer wg.Done()
+			triggerInstance.SignalWorkerDraining()
+		}(triggerInstance, wg)
+	}
```

Comment:
When you iterate on an array of objects and pass the object to a goroutine, you might experience some race conditions where the object is replaced with the next iteration before being used in the goroutine (since it is passed by reference).

A way to workaround it is to redefine the object before invoking the goroutine, e.g:
```suggestion
	for _, triggerInstance := range p.triggers {
	        triggerInstance := triggerInstance
		wg.Add(1)

		// drains all workers in trigger (for each trigger in parallel)
		go func(triggerInstance trigger.Trigger, wg *sync.WaitGroup) {
			defer wg.Done()
			triggerInstance.SignalWorkerDraining()
		}(triggerInstance, wg)
	}
```

Reply:
as discussed, this is not an issue in case of passing parameter directly to function